### DISCUSSION TITLE
getting_started/first_3d_game/02.player_input

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#109] [User-contributed notes]
# getting_started/first_3d_game/02.player_input

In the next two lessons, we will design the player scene, register custom input actions, and code player movement. By the end, you'll have a playable character that moves in eight directions. Creat...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/02.player_input.html

**Comment by bandinopla** [comment]
My character for some reasons ends up hovering the floor by a lot but the collider is set as instructed... why could this be?

**Reply by bandinopla** [reply]
found the issue, it was my mistake, when they said to move the floor down to see the grid, I selected the floor in the viewport, which selected the box mesh, not the Node3D Ground, so the mesh looked 2 units down in the Y but everything else was at 0 so the player was hitting the collider at 0 making it look like it was floating... so I moved the mesh back to 0 and then moved the main ground node 2 units down... :D

**Comment by noikoinoi** [comment]
I'm still new to godot and game development in general, but can someone tell me why this tutorial uses the Pivot(Node3D) node?
I've noticed that I achieve the same result by setting the direction on the Character (Node3D) node.

**Reply by bandinopla** [reply]
it is usually good to wrap things in "empty" objects to have better control over them or add more things in the future, etc... if you think in abstract terms, the "charater" is an empty thing and you give life to it by filling it with stuff... Like, you are not just your bones, you are your organs too etc... so you in 3D starts with an empty Node3D and fill in the rest. We are all empty 3d nodes at the begining...

**Reply by v-kydela** [reply]
It's for rotating the character. If you rotate the glb directly then it might not rotate around the point you want, and if you rotate the player node then everything else inside the player will rotate too. For this particular game it may not be an issue, but imagine you had something like a life bar attached to the player that you didn't want rotating. It's even common to have the camera attached to the player.

**Comment by asperkraken** [comment]
Note: Blender must be installed before you can import the files. Also the extension is now .blend.

In addition, the imported model is washed out. I had to click it to "editable children" and move the light. Working in Compatibility mode.

**Reply by asperkraken** [reply]
Second note: This is blender 4.2 version, I noticed the editor mentions Blender 3. However, not sure this is it - the Mob model is just fine on import. Maybe the player model just needs some tweaking due to version changes.

**Reply by Calinou** [reply]
Blender is not required, as the demo project already includes glTF files which Godot can import directly. You can disable .blend import if you see the Blender import dialog pop up. (Blender import can be disabled in the project settings if you ant to get rid of the dialog for this project.)

**Comment by potentialnull** [comment]
I think it would be nice to instruct on how to make a simple prototype player model instead of (or as an alternative to) using a pre-made asset.

**Reply by Calinou** [reply]
Making a 3D asset is out of scope for a game engine tutorial, as it requires external software. It's also a rather complicated process as you'd also need to create animations for it, which is notoriously time-consuming.

**Reply by potentialnull** [reply]
It actually doesn't require external software because the prototype character doesn't need to be anything special. You can make a CollisionShape3D and MeshInstance3D as children of a CharacterBody3D. I made mine a capsule following the "Learn Godot by Making Minecraft" video on youtube, and shaded it #468CBF to give it some pizzazz. I'll call it Captain Capsule and I will love it because it is mine and I made it.

Mr Lovato already covers animation in part 9 of this tutorial.

**Comment by ComposerLOST** [comment]
In the Input Map, I don't have a "Manual Input" tab--it shows the manual menu under the "listening for input" and "filter inputs" boxes so you can still select the axes, though. (v4.3.stable.official)

**Comment by RedBreadFrog** [comment]
Why use a sphere over, say, a Cylinder?  Is there a strong reason that a sphere is preferred?  Or does it not matter too much for this tutorial, and you'd use whatever worked best in future projects?

**Reply by Justin-Archmagus** [reply]
I wondered this too, I agree with your insinuation that a cylinder is probably a better match for the model shape unless we will be learning something about the benefits of spheres later.

**Reply by Calinou** [reply]
Sphere collision is faster and has fewer bugs than cylinder collision (when using GodotPhysics, that is).

JoltPhysics in 4.4 should deal with cylinder collision much more reliably, but it's still slower than sphere collision.

**Reply by Justin-Archmagus** [reply]
Good to know, thanks!  The way the sphere collision was used to figure out whether the player was above the mob led to some logic bugs depending on the angle of the collision normal but that can be fixed with either placement tweaking or changing the relative height detection logic.

**Comment by tammai** [comment]
How to map touch screen actions like swipe and tap?

**Reply by nightgaspard** [reply]
I would look into [InpuEventScreenTouch class](https://docs.godotengine.org/en/stable/classes/class_inputeventscreentouch.html) there's also probably independent tutorials like youtube videos and tutorial articles covering this on the web.

**Comment by matyasgyimesi** [comment]
Dragging player.glb onto Pivot doesn't work. I'm using Godot 4.3 in Windows 11. I have the node selceted and try to drag onto it and nothing happens.

**Reply by matyasgyimesi** [reply]
Restarting Godot fixed the issue

**Comment by oguzun1** [comment]
E 0:00:16:496   player.gd:25 @ _physics_process(): Node not found: "Pivot" (relative to "/root/main/player").
 ??

**Comment by v-kydela** [comment]
"We could use these here, but instead we're defining our own to support gamepads."

The built-in actions do support gamepads, just like the joypad inputs we defined.

The tedium of adding every input for every action manually without any axes or shortcuts really isn't selling this game engine.

**Reply by Calinou** [reply]
While you *can* use built-in `ui_*` actions for gameplay logic, it's good practice to use your own actions for gameplay. The reason is that `ui_*` actions are meant to be used for UI interactions only. For instance, if you support input remapping, it should affect the gameplay but not affect UI interactions.

**Reply by v-kydela** [reply]
That's fine, but is there a faster way to define actions for arrow keys? Can the `ui_*` actions be copied and pasted?

**Reply by Calinou** [reply]
>  Can the `ui_*` actions be copied and pasted?

This isn't possible currently. That said, there's a proposal for exporting the input map from a project to another, so you can create quickly reusable presets:

- https://github.com/godotengine/godot-proposals/issues/423

**Reply by FrackinFamous** [reply]
> That's fine, but is there a faster way to define actions for arrow keys? Can the `ui_*` actions be copied and pasted?

You can add things to the `ui_*` actions like WASD for directions. I like this in menus anyway. As Calinou said though, we should not use those for in game logic. The docs go into more depth on this in input section. There I believe 7 levels of checking for input. Anyway, you can script your keybinds you could even make the script using class_name and then drag it onto any "player" you want as a composable node.

**Reply by tiham93** [reply]
You can just copy the [input] section in your project.godot root file over, or put it in a snippet for your text editor. It's what I like about Godot: it's all plain text so you can do whatever you want.

**Comment by gruszczy** [comment]
The tutorial recommends moving the collision shape to align with the grid, but the character mesh remains in the original place. This makes shape not really align with it. Is that intended? If yes, maybe worth mentioning in the tutorial around the section that suggests moving the shape up.

**Reply by alexpalade** [reply]
I think I got the idea. The bottom of the sphere is where the character will rest on the floor. By default, the sphere extends below the character model which makes it look like it's floating. Move the collision sphere until the bottom aligns with the bottom of the model. It's easier to position when viewed from the ortographic X axis view (click on the axes gizmo).

**Reply by FrackinFamous** [reply]
What I believe he means is they say not to touch the player models origin but we move the CollisionMesh3D up so the edge of it's radius touches the floor. When we turn visibility back on if done exactly as pictured and explained, the mesh looks like it's a ball sitting on 0y while the models mesh is still splitting 0y with it's center. It is very jarring but meant to be simplistic for this tutorial. Which if you like to use your brain in tutorials to figure out "why" this is jarring. Normally the collision mesh would be wrap the character better but they did not want to mess with having a non 0y origin character model for this scope. The reason they move the collider up is to make sure the physics engine is not trying to constantly eject if from the floor.

**Comment by gitoli56** [comment]
The node hierarchy Player (Node3D) -> Pivot (Node) -> Character (Node3D) is coming out of the blue for a beginner. It would help to provide some explanation about why things are organized/decoupled this way and whether there are alternatives.
